File: backend/banner_outcomes.py
"""Conservative featured and guarantee inference from documented pool rules.

Derivation uses complete history, never filtered display rows. A catalog rarity
or an unrecognized item alone cannot establish that a reward was featured.
"""
from collections import defaultdict
import json
from pathlib import Path


def load_banner_rules():
    path = Path(__file__).with_name("banner_rules.json")
    return json.loads(path.read_text(encoding="utf-8"))


def annotate_banner_outcomes(rows, endpoint_host, rules=None):
    """Mutate newest-first rows after coverage annotation, before filtering.

    Source order at equal timestamps is already newest-first. Reverse it along
    with timestamps, preserving the source's ordering of individual rewards.
    Initial guarantee is unknown even for history beginning on launch day.
    """
    rules = load_banner_rules() if rules is None else rules
    provider_known = endpoint_host in rules["hosts"]
    pools = defaultdict(list)
    for pool in rules["pools"]:
        pools[(pool["type_id"], pool["pool_id"])].append(pool)
    fixed_loss_pools = defaultdict(list)
    for pool in rules.get("fixed_loss_pools", []):
        fixed_loss_pools[pool["type_id"]].append(pool)
    by_type = defaultdict(list)
    for row in rows:
        by_type[row["type_id"]].append(row)
    for type_id, newest_first in by_type.items():
        guarantee = None
        interval = None
        uncertainty = "unknown_start"
        chronological = sorted(newest_first, key=lambda row: (
            row["timestamp"], -row.get("timestamp_order", 0), -row.get("id", 0)))
        for row in chronological:
            result = dict(featured=None, outcome="not_applicable", guarantee_before=None,
                          guarantee_after=None, featured_pity=None, reason=None)
            row["banner_result"] = result
            if type_id not in (3, 4):
                result["reason"] = "unsupported_type"
                continue
            if row.get("gap_before") or row["rarity"] == "Unknown":
                guarantee, interval = None, None
                uncertainty = "history_gap" if row.get("gap_before") else "unknown_item"
            matching = [pool for pool in pools[(type_id, row["pool_id"])]
                        if (not pool.get("start") or row["timestamp"] >= pool["start"])
                        and (not pool.get("end") or row["timestamp"] < pool["end"])]
            pool = matching[0] if len(matching) == 1 else None
            fixed_candidates = [candidate for candidate in fixed_loss_pools[type_id]
                                if (not candidate.get("start") or row["timestamp"] >= candidate["start"])
                                and (not candidate.get("end") or row["timestamp"] < candidate["end"])]
            fixed_pool = fixed_candidates[0] if len(fixed_candidates) == 1 else None
            if pool is None and fixed_pool is not None and any(
                    item in fixed_pool["item_ids"]
                    for candidate in pools[(type_id, row["pool_id"])]
                    for item in candidate["featured"]):
                # A rerun of a standard Elite needs exact dates to distinguish
                # featured rewards from the same item appearing as a loss.
                fixed_pool = None
            classification_pool = pool if pool is not None else fixed_pool
            unknown = ("unknown_provider" if not provider_known else
                       "unknown_pool" if classification_pool is None else None)
            if unknown:
                guarantee, interval, uncertainty = None, None, unknown
                result.update(outcome="unknown" if row["rarity"] == "Elite" else "not_applicable",
                              reason=unknown)
                continue
            if interval is not None:
                interval += 1
            result.update(guarantee_before=guarantee, guarantee_after=guarantee)
            if row["rarity"] != "Elite":
                if row["rarity"] == "Unknown":
                    result["reason"] = uncertainty
                continue
            off_banner = row.get("kind") == classification_pool["kind"]
            # A documented fixed loss roster also identifies featured Elites when
            # historical banner dates are unavailable. Exact dated pools win.
            featured = off_banner and (row["item_id"] in pool["featured"] if pool is not None
                                       else row["item_id"] not in fixed_pool["item_ids"])
            if not featured and not off_banner:
                # A catalog unknown or mismatched reward kind cannot prove a loss.
                guarantee, interval, uncertainty = None, None, "unknown_item"
                result.update(outcome="unknown", guarantee_after=None, reason=uncertainty)
            elif featured:
                result.update(featured=True, outcome="guaranteed" if guarantee is True else
                              "win" if guarantee is False else "unknown",
                              guarantee_after=False, featured_pity=interval,
                              reason=uncertainty if guarantee is None else None)
                guarantee, interval, uncertainty = False, 0, None
            elif guarantee is True:
                result.update(featured=False, outcome="unknown", guarantee_after=True,
                              reason="guarantee_conflict")
                guarantee, interval, uncertainty = True, None, None
            else:
                # A known off-banner Elite proves this draw was not guaranteed.
                result.update(featured=False, outcome="loss", guarantee_before=False,
                              guarantee_after=True, reason=None)
                guarantee, uncertainty = True, None
    return rows


def summarize_banner_outcomes(rows):
    """Summarize a complete newest-first recruitment type, independently of UI filters."""
    elites = [row["banner_result"] for row in rows
              if row["rarity"] == "Elite" and row["type_id"] in (3, 4) and row.get("banner_result")]
    latest = rows[0].get("banner_result") if rows else None
    return dict(featured_count=sum(result["featured"] is True for result in elites),
                off_banner_count=sum(result["featured"] is False for result in elites),
                wins=sum(result["outcome"] == "win" for result in elites),
                losses=sum(result["outcome"] == "loss" for result in elites),
                guaranteed=sum(result["outcome"] == "guaranteed" for result in elites),
                unknown_elites=sum(result["featured"] is None for result in elites),
                unknown_outcomes=sum(result["outcome"] == "unknown" for result in elites),
                featured_intervals=[result["featured_pity"] for result in elites
                                    if result["featured"] is True and result["featured_pity"] is not None],
                current_guarantee=latest["guarantee_after"] if latest else None)

File: web/src/lib/banner-outcomes.ts
import rules from '../../../backend/banner_rules.json' with { type: 'json' };
import type { Pull } from './api.ts';

export interface BannerRule {
  type_id: number;
  pool_id: number;
  featured: number[];
  kind: string;
  start?: string;
  end?: string;
}
export interface BannerRules {
  hosts: string[];
  pools: BannerRule[];
  fixed_loss_pools?: {
    type_id: number;
    kind: string;
    item_ids: number[];
    start?: string;
    end?: string;
  }[];
}
export interface BannerResult {
  featured: boolean | null;
  outcome: 'win' | 'loss' | 'guaranteed' | 'unknown' | 'not_applicable';
  guarantee_before: boolean | null;
  guarantee_after: boolean | null;
  featured_pity: number | null;
  reason:
    | 'unknown_pool'
    | 'unknown_item'
    | 'unknown_start'
    | 'history_gap'
    | 'guarantee_conflict'
    | 'unsupported_type'
    | 'unknown_provider'
    | null;
}
export interface FeaturedSummary {
  featured_count: number;
  off_banner_count: number;
  wins: number;
  losses: number;
  guaranteed: number;
  unknown_elites: number;
  unknown_outcomes: number;
  featured_intervals: number[];
  current_guarantee: boolean | null;
}

/** Derive once from complete, coverage-annotated history, never a filtered page.
 * Imported outcome flags are deliberately ignored: immutable records and the
 * reviewed pool catalog are the evidence, while gaps invalidate carry state.
 */
export function annotateBannerOutcomes<T extends Pull>(
  rows: T[],
  endpointHost: string | null,
  catalog: BannerRules = rules
): void {
  const knownProvider = endpointHost !== null && catalog.hosts.includes(endpointHost);
  const lookup = new Map<string, BannerRule[]>();
  for (const rule of catalog.pools) {
    const key = `${rule.type_id}/${rule.pool_id}`;
    lookup.set(key, [...(lookup.get(key) || []), rule]);
  }
  const groups = new Map<number, T[]>();
  for (const row of rows) {
    if (!groups.has(row.type_id)) groups.set(row.type_id, []);
    groups.get(row.type_id)!.push(row);
  }
  for (const [type, group] of groups) {
    let guarantee: boolean | null = null;
    let featuredPity: number | null = null;
    let uncertainty: BannerResult['reason'] = 'unknown_start';
    // Stable source order is newest first even within a single timestamp.
    const chronological = [...group].sort(
      (a, b) =>
        a.timestamp.localeCompare(b.timestamp) ||
        b.timestamp_order - a.timestamp_order ||
        b.id - a.id
    );
    for (const row of chronological) {
      const result: BannerResult = {
        featured: null,
        outcome: 'not_applicable',
        guarantee_before: guarantee,
        guarantee_after: guarantee,
        featured_pity: null,
        reason: null
      };
      row.banner_result = result;
      if (type !== 3 && type !== 4) {
        result.reason = 'unsupported_type';
        continue;
      }
      if (row.gap_before || row.rarity === 'Unknown') {
        guarantee = null;
        featuredPity = null;
        uncertainty = row.gap_before ? 'history_gap' : 'unknown_item';
      }
      const matching = (lookup.get(`${type}/${row.pool_id}`) || []).filter(
        (rule) =>
          (!rule.start || row.timestamp >= rule.start) && (!rule.end || row.timestamp < rule.end)
      );
      const rule = knownProvider && matching.length === 1 ? matching[0] : null;
      const fixedMatches = (catalog.fixed_loss_pools ?? []).filter(
        (pool) =>
          pool.type_id === type &&
          (!pool.start || row.timestamp >= pool.start) &&
          (!pool.end || row.timestamp < pool.end)
      );
      const candidate = fixedMatches.length === 1 ? fixedMatches[0] : null;
      // Standard dolls have also had dedicated rate-ups. Outside their mapped
      // dates, membership in the loss roster alone cannot classify that pool.
      const standardRateUp =
        candidate &&
        (lookup.get(`${type}/${row.pool_id}`) ?? []).some((pool) =>
          pool.featured.some((id) => candidate.item_ids.includes(id))
        );
      const fixed = knownProvider && !standardRateUp ? candidate : null;
      if (!rule && !fixed) {
        guarantee = null;
        featuredPity = null;
        uncertainty = knownProvider ? 'unknown_pool' : 'unknown_provider';
      }
      if (featuredPity !== null) featuredPity++;
      result.guarantee_before = guarantee;
      if (row.rarity === 'Elite') {
        result.outcome = 'unknown';
        // Targeted banners have a known loss roster even when their dated
        // featured mapping is missing. Explicit banner mappings take precedence.
        const kind = rule?.kind ?? fixed?.kind;
        const featured = rule
          ? rule.featured.includes(row.item_id)
          : fixed && !fixed.item_ids.includes(row.item_id);
        if (kind && row.kind === kind && featured) {
          result.featured = true;
          result.outcome =
            guarantee === true ? 'guaranteed' : guarantee === false ? 'win' : 'unknown';
          result.reason = guarantee === null ? uncertainty : null;
          result.featured_pity = featuredPity;
          guarantee = false;
          featuredPity = 0;
          uncertainty = null;
        } else if (kind && row.kind === kind) {
          result.featured = false;
          if (guarantee === true) {
            result.reason = 'guarantee_conflict';
            featuredPity = null;
          } else {
            // An off-banner result itself proves that this roll was not guaranteed.
            result.guarantee_before = false;
            result.outcome = 'loss';
          }
          guarantee = true;
          uncertainty = null;
        } else {
          guarantee = null;
          featuredPity = null;
          uncertainty = rule || fixed ? 'unknown_item' : uncertainty;
          result.reason = uncertainty;
        }
      } else if ((!rule && !fixed) || row.rarity === 'Unknown') {
        result.reason = uncertainty;
      }
      result.guarantee_after = guarantee;
    }
  }
}

/** Summary stays independent of rarity selection and pagination. */
export function summarizeBannerOutcomes(rows: Pull[]): FeaturedSummary {
  const summary: FeaturedSummary = {
    featured_count: 0,
    off_banner_count: 0,
    wins: 0,
    losses: 0,
    guaranteed: 0,
    unknown_elites: 0,
    unknown_outcomes: 0,
    featured_intervals: [],
    current_guarantee: rows[0]?.banner_result?.guarantee_after ?? null
  };
  for (const row of rows) {
    if (row.rarity !== 'Elite' || (row.type_id !== 3 && row.type_id !== 4)) continue;
    const result = row.banner_result;
    if (!result || result.outcome === 'not_applicable') continue;
    if (result?.featured === true) summary.featured_count++;
    else if (result?.featured === false) summary.off_banner_count++;
    else summary.unknown_elites++;
    if (result?.outcome === 'win') summary.wins++;
    else if (result?.outcome === 'loss') summary.losses++;
    else if (result?.outcome === 'guaranteed') summary.guaranteed++;
    else summary.unknown_outcomes++;
    if (result?.featured_pity != null) summary.featured_intervals.push(result.featured_pity);
  }
  return summary;
}
